src/stan-users-guide/parallel-computing.Rmd

```diff
@@ -1,12 +1,252 @@
+
+The ```reduce_sum``` call:
+
+```
+real sum = reduce_sum(func, x, grainsize, s1, s2, ...)
+```
+
+can be replaced by either:
+
+```
+real sum = func(1, size(x), x, s1, s2, ...)
+```
+
+or the code:
+
+```
+real sum = 0.0;
+for(i in 1:size(x)) {
+  sum = sum + func(i, i, { x[i] }, s1, s2, ...);
+}
+```
+
+### Example: Logistic Regression
+
+Logistic Regression is a useful example to clarify both the syntax
+and semantics of `reduce_sum` and how it can be used to speed up a typical
+model.
+
+A basic logistic regression can be coded in Stan as:
+
+```
+data {
+  int N;
+  int y[N];
+  vector[N] x;
+}
+parameters {
+  vector[2] beta;
+}
+model {
+  beta ~ std_normal();
+  y ~ bernoulli_logit(beta[1] + beta[2] * x);
+}
+```
+
+In this model predictions are made about the `N` outputs `y` using the
+covariate `x`. The intercept and slope of the linear equation are to be estimated.
+
+The key point to getting this calculation into `reduce_sum`, is recognizing that
+the statement:
+
+```
+y ~ bernoulli_logit(beta[1] + beta[2] * x);
+```
+
+can be rewritten (up to a proportionality constant) as:
+```
+for(n in 1:N) {
+  target += bernoulli_logit_lpmf(y[n] | beta[1] + beta[2] * x[n])
+}
+```
+
+Now it is clear that the calculation is the sum of a number of conditionally
+independent Bernoulli log probability statements, which is the condition where
+`reduce_sum` is useful.
+
+To use `reduce_sum`, a function must be written that can be used to compute
+arbitrary partial sums of the total sum.
+
+Using the interface defined in [Reduce-Sum](#reduce-sum), such a function
+can be written like:
+
+```
+functions {
+  real partial_sum(int start, int end,
+                   int[] y_subset,
+                   vector x,
+                   vector beta) {
+    return bernoulli_logit_lpmf(y_subset | beta[1] + beta[2] * x[start:end]);
+  }
+}
+```
+
+And the likelihood statement in the model can now be written:
+
+```
+target += partial_sum(1, N, y, x, beta); // Sum terms 1 to N in the likelihood
+```
+
+In this example, `y` was chosen to be sliced over because there
+is one term in the summation per value of `y`. Technically `x` would  have
+worked as well. Use whatever conceptually makes the most sense.
+
+Because `x` is a shared argument, it is subset manually with `start:end`.
+
+With this function, `reduce_sum` can be used to automatically parallelize the
+likelihood:
+
+```
+int grainsize = 100;
+target += reduce_sum(partial_sum, y,
+                     grainsize,
+                     x, beta);
+```
+
+`reduce_sum` automatically breaks the sum into roughly `grainsize` sized pieces
+and computes them in parallel. `grainsize = 1` specifies that the grainsize should
```

Comment:
0 or 1?  below you say `grainsize = 0` - shouldn't `grainsize = 1` specify no parallelization?

Reply:
Sorry, had a problem with the text. ```1``` leaves the parallelization up to TBB (it can make slices as small as 1 is what that basically means)